kong/db/strategies/postgres/connector.lua

```diff
@@ -371,9 +372,46 @@ function _mt:setkeepalive()
 end
 
 
+function _mt:acquire_lock()
+  if get_phase() == "init" or get_phase() == "init_worker" then
```

Comment:
[wait()](https://github.com/openresty/lua-resty-core/blob/master/lib/ngx/semaphore.md#wait) can only be called from contexts that can yield, so I think more restrictions should be applied here. Also, better cache the result of `get_phase()`, especially if we add more checks.

Reply:
In generally, the caller (`:query`) can also only be called in functions that yield, or init/init_worker. Given that we don't make restrictions in which phases `:query` can be called, is a separate check needed here as well? And, should that check live here, or in `:query`? I'd be happy to make a separate patchset to do some more phase sanity checking in `:query`, that sounds like a good defensive step.

Reply:
The paradox here comes from the fact that `:query` is compatible in non-yieldable contexts (header_filter, body_filter, log...) with the magical fallback to LuaSocket. In those cases, the call to this semaphore will throw a Lua error, which is what I was concerned about and was asking for a way to cover those cases, for the programmer's sake (it is not a mistake to call this semaphore in those phases if `:query` is supported in those phases).

However, `:query` should _not_ be supported in those phases. There is a pending work item which is to *unload* LuaSocket once initialization has completed, in order to ensure that we can never block the NGINX runloop unknowingly (better fail hard and early than letting such a fallback ship to users and destroy performance under unknown cases. E.g. in a plugin with a particular configuration or other such twisted cases that make it hard to catch in pre-release benchmarking and smoke testing).

Reply:
@thibaultcha do you still want this restriction to be tightened prior to merging? Otherwise I think this is good to go.

Reply:
I'd say it is fine for now, at least a minor concern in comparison to the HTTP 500 error we return to clients (vs. a more appropriate 503 status code imho). What do you think?

Reply:
> What do you think?

I agree with the current state of the PR in which it does not add the check in this function, as it is only called by `:query` and the real issue is where can `:query` be called. If I were to suggest anything, it would be to make the acquire and release functions `local function`s to the module passing an explicit `self` (instead of having them `_mt:` functions) to make their private status more evident. But this is not a blocker.